numpy/_core/src/umath/string_fastsearch.h

```diff
@@ -222,35 +424,67 @@ rfindchar(CheckedIndexer<char_type> s, Py_ssize_t n, char_type ch)
     return -1;
 }
 
-
-/* Change to a 1 to see logging comments walk through the algorithm. */
+#undef MEMCHR_CUT_OFF
+
+/**
+ * @file_internal
+ * @brief Conditional logging for string fast search.
+ *
+ * Set to 1 to enable logging macros.
+ *
+ * @note These macros are used internally for debugging purposes
+ * and will be undefined later in the code.
+ */
 #if 0 && STRINGLIB_SIZEOF_CHAR == 1
-# define LOG(...) printf(__VA_ARGS__)
-# define LOG_STRING(s, n) printf("\"%.*s\"", (int)(n), s)
-# define LOG_LINEUP() do {                                         \
+/** Logs formatted output. */
+#define LOG(...) printf(__VA_ARGS__)
+
+/** Logs a string with a given length. */
+#define LOG_STRING(s, n) printf("\"%.*s\"", (int)(n), s)
+
+/** Logs the current state of the algorithm. */
+#define LOG_LINEUP() do {                                          \
     LOG("> "); LOG_STRING(haystack, len_haystack); LOG("\n> ");    \
     LOG("%*s",(int)(window_last - haystack + 1 - len_needle), ""); \
     LOG_STRING(needle, len_needle); LOG("\n");                     \
 } while(0)
 #else
-# define LOG(...)
-# define LOG_STRING(s, n)
-# define LOG_LINEUP()
+#define LOG(...)
+#define LOG_STRING(s, n)
+#define LOG_LINEUP()
 #endif
 
+/**
+ * @file_internal
+ * @brief Perform a lexicographic search for the maximal suffix in
+ * a given string.
+ *
+ * This function searches through the `needle` string to find the
+ * maximal suffix, which is essentially the largest lexicographic suffix.
+ * Essentially this:
+ * - max(needle[i:] for i in range(len(needle)+1))
+ *
+ * Additionally, it computes the period of the right half of the string.
+ *
+ * @param needle The string to search in.
+ * @param len_needle The length of the needle string.
+ * @param return_period Pointer to store the period of the found suffix.
+ * @param invert_alphabet Flag to invert the comparison logic.
+ * @return The index of the maximal suffix found in the needle string.
+ *
+ * @note If `invert_alphabet` is non-zero, character comparisons are reversed,
+ * treating smaller characters as larger.
+ *
+ */
 template <typename char_type>
 static inline Py_ssize_t
-_lex_search(CheckedIndexer<char_type> needle, Py_ssize_t len_needle,
```

Comment:
deleting a single leading underscore like this is fine, since you can still meaningfully grep CPython for `lex_search`.